cvat-core/src/annotations-objects.ts

```diff
@@ -2228,10 +2235,39 @@ export class MaskShape extends Shape {
             }
         }
 
-        for (const object of Object.values(updatedObjects)) {
-            object.points = mask2Rle(masks[object.clientID]);
-            object.updated = Date.now();
-        }
+        let emptyMaskOccured = false;
+        let stashedPoints = [];
+        const undo = (): void => {
+            for (const [index, object] of Object.values(updatedObjects).entries()) {
+                if (object.points.length < 2) {
+                    object.removed = false;
+                }
+                object.points = stashedPoints[index];
+                object.updated = Date.now();
+            }
+        };
+
+        const redo = (): void => {
+            stashedPoints = [];
+            for (const object of Object.values(updatedObjects)) {
+                const points = mask2Rle(masks[object.clientID]);
```

Comment:
There will be a problem with huge memory consumption

In closure it refers to variable `masks` that contains all unpacked masks stored on the frame (imagine that there are 100+ masks per frame).
And undo/redo stores up to 128 actions as far as I remember 

It gigabytes of data

Reply:
`masks` store not all masks on the frame but only affected ones, but still there can be a problem.
Ive added limitations to annotation history, so there only can be `20` undos for removing underlying pixels. (maybe we should change this number to even lower value)